unittests/etl/ForwardingSourceTests.cpp

```diff
@@ -118,3 +119,90 @@ TEST_F(ForwardingSourceOperationsTests, Success)
         EXPECT_EQ(*result, expectedReply) << *result;
     });
 }
+
+struct ForwardingSourceCacheTests : ForwardingSourceOperationsTests {
+    ForwardingSourceCacheTests()
+    {
+        forwardingSource =
+            ForwardingSource{"127.0.0.1", "11114", std::chrono::seconds{100}, std::chrono::milliseconds{1}};
+    }
+};
+
+TEST_F(ForwardingSourceCacheTests, Cache)
+{
+    boost::json::object const request = {{"command", "server_state"}};
+    auto const response = R"({"reply":"some_reply"})";
+
+    boost::asio::spawn(ctx, [&](boost::asio::yield_context yield) {
+        auto connection = serverConnection(yield);
+
+        auto const receivedMessage = connection.receive(yield);
+        [&]() { ASSERT_TRUE(receivedMessage); }();
+        EXPECT_EQ(*receivedMessage, boost::json::serialize(request)) << *receivedMessage;
+
+        {
+            auto const sendError = connection.send(response, yield);
+            [&]() { ASSERT_FALSE(sendError) << *sendError; }();
+        }
+
+        auto const sendError = connection.send("some other message", yield);
+        [&]() { ASSERT_FALSE(sendError) << *sendError; }();
+    });
+
+    runSpawn([&](boost::asio::yield_context yield) {
+        for (int i = 0; i < 4; ++i) {
+            auto result = forwardingSource.forwardToRippled(request, {}, yield);
+            [&]() { ASSERT_TRUE(result); }();
+
+            auto expectedReply = boost::json::parse(response).as_object();
+            expectedReply["forwarded"] = true;
+            EXPECT_EQ(*result, expectedReply) << *result;
+        }
+    });
+}
+
+TEST_F(ForwardingSourceCacheTests, ResponseWithErrorNotCached)
+{
+    boost::json::object const request = {{"command", "server_state"}};
+    auto const errorResponse = R"({"reply":"some_reply","error":"some_error"})";
```

Comment:
nit: Add some spaces to JSONs? 🏅

Reply:
I can't because they are being compared with JSONs generated automatically by `boost::json`.

Reply:
I don't insist but the way we usually do this is by creating a `json::object` from the string and then comparing the objects instead :)